// angular2-spinner/dist/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var angular2_spinner_1 = require("./angular2-spinner");
__export(require("./angular2-spinner"));
var SpinnerModule = SpinnerModule_1 = (function () {
    function SpinnerModule() {
    }
    SpinnerModule.forRoot = function () {
        return {
            ngModule: SpinnerModule_1,
            providers: []
        };
    };
    return SpinnerModule;
}());
SpinnerModule = SpinnerModule_1 = __decorate([
    core_1.NgModule({
        declarations: [
            angular2_spinner_1.SpinnerComponent
        ],
        exports: [
            angular2_spinner_1.SpinnerComponent,
        ],
        schemas: [core_1.NO_ERRORS_SCHEMA],
    })
], SpinnerModule);
exports.SpinnerModule = SpinnerModule;
var SpinnerModule_1;
