// angular2-dynamic-component/src/DynamicComponent.ngfactory.ts
/**
 * @fileoverview This file is generated by the Angular template compiler.
 * Do not edit.
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride}
 */
 /* tslint:disable */


import * as import0 from '@angular/core';
import * as import1 from './DynamicComponent';
import * as import2 from '@angular/http';
import * as import3 from './DynamicCache';
const styles_DynamicComponent:any[] = ([] as any[]);
export const RenderType_DynamicComponent:import0.RendererType2 = import0.ɵcrt({
  encapsulation: 2,
  styles: styles_DynamicComponent,
  data: {}
}
);
export function View_DynamicComponent_0(l:any):import0.ɵViewDefinition {
  return import0.ɵvid(0,([] as any[]),(null as any),(null as any));
}
function View_DynamicComponent_Host_0(l:any):import0.ɵViewDefinition {
  return import0.ɵvid(0,[
    (l()(),import0.ɵeld(16777216,(null as any),(null as any),1,'DynamicComponent',([] as any[]),(null as any),(null as any),(null as any),View_DynamicComponent_0,RenderType_DynamicComponent)),
    import0.ɵdid(704512,(null as any),0,import1.DynamicComponent,[
      'DynamicExtraModules',
      import0.ViewContainerRef,
      import0.Compiler,
      import2.Http,
      import3.DynamicCache
    ]
    ,(null as any),(null as any))
  ]
  ,(null as any),(null as any));
}
export const DynamicComponentNgFactory:import0.ComponentFactory<import1.DynamicComponent> = import0.ɵccf('DynamicComponent',import1.DynamicComponent,View_DynamicComponent_Host_0,{
  componentType: 'componentType',
  componentTemplate: 'componentTemplate',
  componentStyles: 'componentStyles',
  componentContext: 'componentContext',
  componentTemplateUrl: 'componentTemplateUrl',
  componentTemplatePath: 'componentTemplatePath',
  componentDefaultTemplate: 'componentDefaultTemplate',
  componentRemoteTemplateFactory: 'componentRemoteTemplateFactory',
  componentModules: 'componentModules'
}
,{
  dynamicComponentReady: 'dynamicComponentReady',
  dynamicComponentBeforeReady: 'dynamicComponentBeforeReady'
}
,([] as any[]));
