// angular2-spinner/angular2-spinner.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var SpinnerComponent = (function () {
    function SpinnerComponent() {
        this.size = 25;
        this.tickness = 2;
        this.color = '#4f6aa7';
        this.opacity = ".1";
        this.secondColor = '';
    }
    SpinnerComponent.prototype.ngOnInit = function () {
        var c = this.hexToRgb(this.color);
        this.secondColor = 'rgba(' + c.r + ',' + c.g + ',' + c.b + ', ' + this.opacity + ')';
    };
    SpinnerComponent.prototype.hexToRgb = function (hex) {
        var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
        return result ? {
            r: parseInt(result[1], 16),
            g: parseInt(result[2], 16),
            b: parseInt(result[3], 16)
        } : null;
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], SpinnerComponent.prototype, "size", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], SpinnerComponent.prototype, "tickness", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], SpinnerComponent.prototype, "color", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], SpinnerComponent.prototype, "opacity", void 0);
    SpinnerComponent = __decorate([
        core_1.Component({
            moduleId: module.id,
            selector: 'spinner',
            template: '',
            styleUrls: ['angular2-spinner.css'],
            host: {
                '[style.width]': 'size + "px"',
                '[style.height]': 'size + "px"',
                '[style.borderWidth]': 'tickness + "px"',
                '[style.borderTopColor]': 'color',
                '[style.borderBottomColor]': 'secondColor',
                '[style.borderLeftColor]': 'secondColor',
                '[style.borderRightColor]': 'secondColor'
            },
        }), 
        __metadata('design:paramtypes', [])
    ], SpinnerComponent);
    return SpinnerComponent;
}());
exports.SpinnerComponent = SpinnerComponent;
